python3Packages.jupyter-themes: init at 0.20.0

jherland opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jherland committed:
python3Packages.jupyter-themes: init at 0.20.0

Co-authored-by: Ethan Carter Edwards <ethan@ethancedwards.com>

ethancedwards8 commented on pkgs/development/python-modules/jupyter-themes/default.nix:
```diff
@@ -0,0 +1,36 @@
+  ];
+
+  pythonImportsCheck = [ "jupyterthemes" ];
+
```
add pytestCheckHook too, please

jherland replied:
I tried, but [the tests](https://github.com/dunovank/jupyter-themes/blob/master/tests/test_themes.py) are not written using pytest, so pytest finds nothing to test.

I also tried using nose2pytest to modernize the tests, but could not get it to work.

Actually, reading the tests more thoroughly, I don't see how they could possibly run successfully, as there are references to undefined variables (`mf`, `nf`, `tc`) in the first call to `install_themes()`...

infinisil reviewed (commented):
Not sure why, testing this fails:

```
nix-shell -p 'python3.withPackages (p: [ p.jupyter-themes ])' -I nixpkgs=. --run python3
>>> import jupyterthemes
Traceback (most recent call last):
  File "<python-input-0>", line 1, in <module>
    import jupyterthemes
  File "/nix/store/wn1mr9sh3qyb9anhxfr2vjjgziqys9c3-python3-3.13.4-env/lib/python3.13/site-packages/jupyterthemes/__init__.py", line 5, in <module>
    from . import stylefx
  File "/nix/store/wn1mr9sh3qyb9anhxfr2vjjgziqys9c3-python3-3.13.4-env/lib/python3.13/site-packages/jupyterthemes/stylefx.py", line 2, in <module>
    import lesscpy
ModuleNotFoundError: No module named 'lesscpy'
```

jherland committed:
lesscpy: buildPythonApplication -> buildPythonPackage

We are about to add another Python package (jupyter-themes) which
depends on lesscpy, not the lesscpy application, but the lesscpy
Python package. In order to make that package available in the
dependent's Pythons environment, we need to use buildPythonPackage
instead of buildPythonApplication.

See https://github.com/NixOS/nixpkgs/issues/45503 and
https://github.com/NixOS/nixpkgs/commit/b7e0c40d299c for more details.

jherland commented:
Should be fixed in the latest rebase + push. Please have another look, @infinisil.

infinisil reviewed (approved):
Works now!

It appears that the only difference other than the fact that python packages can provide modules, is that they have the python version as the prefix to their derivation name:

https://github.com/NixOS/nixpkgs/blob/9946a599e6f313e4060b5adcf6024c566a2e84e8/pkgs/development/interpreters/python/python-packages-base.nix#L46-L58
